hash.c

```diff
@@ -2936,19 +2982,44 @@ NOINSERT_UPDATE_CALLBACK(hash_aset_str)
 VALUE
 rb_hash_aset(VALUE hash, VALUE key, VALUE val)
 {
-    bool iter_p = hash_iterating_p(hash);
-
     rb_hash_modify(hash);
 
     if (RHASH_TYPE(hash) == &identhash || rb_obj_class(key) != rb_cString) {
-        RHASH_UPDATE_ITER(hash, iter_p, key, hash_aset, val);
+        RHASH_UPDATE(hash, key, hash_aset, val);
     }
     else {
-        RHASH_UPDATE_ITER(hash, iter_p, key, hash_aset_str, val);
+        RHASH_UPDATE(hash, key, hash_aset_str, val);
     }
+
     return val;
 }
 
+VALUE
+rb_hash_update_value(VALUE hash, VALUE key, VALUE val)
+{
+    bool iter_p = hash_iterating_p(hash);
+
+    rb_hash_modify(hash);
+
+    st_update_callback_func *func = (RHASH_TYPE(hash) == &identhash || rb_obj_class(key) != rb_cString)
+        ? UPDATE_CALLBACK(iter_p, hash_update_value)
+        : UPDATE_CALLBACK(iter_p, hash_update_value_str);
+
+    struct update_arg update_info = {
+        .arg = val,
+        .func = func,
+        .hash = hash,
+        .key  = key,
+        .value = val,
+    };
```

Comment:
This is basically the same as the `update_info` in `tbl_update()`, except we need to access it, so we can fish out the `.arg` after `tbl_update()` finished.

That's why I extracted out `tbl2_update()`.